### ISSUE TITLE
Wrong invoice status for sale order lines with negative qty

### ISSUE THREAD

**Issue by gfcapalbo** [odoo/odoo#118365]
Bug first found in https://github.com/OCA/sale-workflow/issues/2443, in which the line with negative QTY 

Impacted versions:
12-16

Steps to reproduce:

create a quotation
add a line with product qty < 0

Current behavior:
invoiced_status is set as "invoiced" wich is not logically sound.

Expected behavior:
invoice_status should be computed correctly for negative QTY

This incongruity in the computation of invoice_status that happens here: 
https://github.com/odoo/odoo/blob/5bbbec5a80c2d16219c500aa27d26b7166fb6f7c/addons/sale/models/sale_order_line.py#L821-L822

will cause the invoice to be blocked in many operations because technically it has an "invoiced" line, wich is also a line that can never be "non invoiced"

**Comment by vava-odoo** [comment]
Hi @gfcapalbo 
Thanks for the report. Can you explain the business flow behind adding a sale order line with a negative product quantity?
Regards,

**Comment by gfcapalbo** [comment]
@vava-odoo i refer to the explanation in the issue: 

https://github.com/OCA/sale-workflow/issues/2443#issuecomment-1506580269

while it seems a logical impossibility for physical products, sometimes for services id does make sense.
It is used commonly, because explicitly allowed by the core sales datamodel
As i looked into sales module i see that the entire logic of invoiced_status does not consider negative qty in SOL a possibility.
But it is and happens quite often when dealing with services. so i think there needs to be patch to manage correctly qty<0, so that invoced_status should never give "fake invoiced status".

**Comment by vava-odoo** [comment]
Hi @gfcapalbo 
Thanks for sharing this. I shared the feedback internally. Let's see what comes out of this.
What I can say is that this part is probably too central and too deep inside major flows that it will probably not be changed in stable versions. 
Reading the reasons you use this (mainly for traceability), I wonder if there aren't proper standard fluxes for this. For a discount/refund, better directly make a discount/refund, no?
It seems to me you're using this as a feature, but this is not meant for that. There are flows that require negative amount (in Logistics for example), but, correct me if I'm wrong, the ones you mention do not.
Regards

**Comment by gfcapalbo** [comment]
> Hi @gfcapalbo Thanks for sharing this. I shared the feedback internally. Let's see what comes out of this. What I can say is that this part is probably too central and too deep inside major flows that it will probably not be changed in stable versions. Reading the reasons you use this (mainly for traceability), I wonder if there aren't proper standard fluxes for this. For a discount/refund, better directly make a discount/refund, no? It seems to me you're using this as a feature, but this is not meant for that. There are flows that require negative amount (in Logistics for example), but, correct me if I'm wrong, the ones you mention do not. Regards

I agree, it was not intended , but it was allowed, and several third party modules use this possibility and create this situation.

the solution could be in the form :
        - fully supporting qty<0, or 
        - disallowing explicitly qty<0 therefore breaking all third party modules.
      
I realize applying  solution 1  to all stable versions would be problematic, but a clarification of qty<0 could be important for future versions as well.
